specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,248 @@
+from astropy.modeling.models import Linear1D, Gaussian1D
+from astropy.modeling.fitting import LMLSQFitter
+import astropy.units as u
+from functools import cached_property
+from gwcs import wcs
+from gwcs import coordinate_frames as cf
+import numpy as np
+from specutils import SpectralRegion, Spectrum1D
+from specutils.fitting import fit_lines
+
+
+class CalibrationLine():
+
+    def __init__(self, input_spectrum, wavelength, pixel, refinement_method=None,
+                 refinement_kwargs={}):
+        """
+        input_spectrum: `~specutils.Spectrum1D`
+            A one-dimensional Spectrum1D calibration spectrum from an arc lamp or similar.
+        wavelength: `astropy.units.Quantity`
+            The rest wavelength of the line.
+        pixel: int
+            Initial guess of the integer pixel location of the line center in the spectrum.
+        refinement_method: str, optional
+            None: use the actual provided pixel values as anchors for the fit without refining
+            their location.
+            'gradient': Simple gradient descent/ascent to nearest min/max value. Direction
+            defined by ``direction`` in refinement_kwargs.
+            'min'/'max': find min/max within pixel range provided by ``range`` kwarg.
+            'gaussian': fit gaussian to spectrum within pixel range provided by ``range`` kwarg.
+        refinement_kwargs: dict, optional
+            Keywords to set the parameters for line location refinement. Distance on either side
+            of line center to include on either side of gaussian fit is set by int ``range``.
+            Gradient descent/ascent is determined by setting ``direction`` to 'min' or 'max'.
+        """
+        self.input_spectrum = input_spectrum
+        self.wavelength = wavelength
+        self.pixel = pixel
+        self.refinement_method = refinement_method
+        self.refinement_kwargs = refinement_kwargs
+
+        if self.refinement_method in ("gaussian", "min", "max"):
+            if 'range' not in self.refinement_kwargs:
+                raise ValueError(f"You must define 'range' in refinement_kwargs to use "
+                                 f"{self.refinement_method} refinement.")
+        elif self.refinement_method == "gradient" and 'direction' not in self.refinement_kwargs:
+            raise ValueError("You must define 'direction' in refinement_kwargs to use "
+                             "gradient refinement")
```

Comment:
do we want defaults for these options or is it too hard to make reasonable defaults that can be generalized?

Reply:
I would be lean toward making gaussian the default, personally.

Reply:
The issue is that the `range` parameter depends heavily on the properties of your spectrum/array, I think. Hence why I left the default as no refinement.

Reply:
Sorry, I meant automatic defaults for `refinement_kwargs`, definitely `refinement_method`-dependent, and perhaps even estimated from the `input_spectrum` (not sure how complicated that would get - definitely could be a follow-up effort... I just feel like users are constantly going to be seeing this error message).